service/history/historyEventNotifier.go

```diff
@@ -89,92 +98,86 @@ func newHistoryEventNotifier(metrics metrics.Client, workflowIDToShardID func(st
 
 		workflowIDToShardID: workflowIDToShardID,
 
-		shardLocks:         make(map[int]*sync.RWMutex),
-		shardEventsPubsubs: make(map[int]map[workflowIdentifier]map[string]chan *historyEventNotification),
+		eventsPubsubs: collection.NewShardedConcurrentTxMap(1024, hashFn),
 	}
 }
 
 func (notifier *historyEventNotifierImpl) WatchHistoryEvent(
 	identifier *workflowIdentifier) (string, chan *historyEventNotification, error) {
 
-	shardID := notifier.workflowIDToShardID(identifier.workflowID)
-	lock, eventsPubsubs := notifier.createOrGetShards(shardID)
-	lock.Lock()
-	defer lock.Unlock()
-
-	subscribers, ok := eventsPubsubs[*identifier]
-	if !ok {
-		subscribers = make(map[string]chan *historyEventNotification)
-		eventsPubsubs[*identifier] = subscribers
+	channel := make(chan *historyEventNotification, 1)
+	subscriberID := uuid.NewUUID().String()
+	subscribers := map[string]chan *historyEventNotification{
```

Comment:
Do you think it makes sense to call GetAndDo before calling put to prevent unnecessary creation of subscribers map?

Reply:
this depends,

1. calling GetAndDo do subscribe, if subscriber map not found, call PutOrDo
2. allocate the subscriber map and call PutOrDo

if most of time the subscriber map exists, 1 will be more efficient.
if most of time subscriber map does not exists, 2 will be more efficient.

if GC is not a problem, 2 will be a better choice.